File: examples/Python/udplib.py
# A Simple UDP class

import socket

class UDP(object):
    """simple UDP ping class"""
    handle = None   # Socket for send/recv
    port = 0        # UDP port we work on
    address = ''    # Own address
    broadcast = ''  # Broadcast address

    def __init__(self, port, address=None, broadcast=None):
        if address is None:
            local_addrs = socket.gethostbyname_ex(socket.gethostname())[-1]
            for addr in local_addrs:
                if not addr.startswith('127'):
                    address = addr
        if broadcast is None:
            broadcast = '255.255.255.255'

        self.address = address
        self.broadcast = broadcast
        self.port = port
        # Create UDP socket
        self.handle = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)

        # Ask operating system to let us do broadcasts from socket
        self.handle.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)

        # Bind UDP socket to local port so we can receive pings
        self.handle.bind(('', port))

    def send(self, buf):
        self.handle.sendto(buf, 0, (self.broadcast, self.port))

    def recv(self, n):
        buf, addrinfo = self.handle.recvfrom(n)
        if addrinfo[0] != self.address:
            print("Found peer %s:%d" % addrinfo)
        return buf

File: examples/Java/Udplib.java
import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.SocketAddress;
import java.net.StandardSocketOptions;
import java.nio.ByteBuffer;
import java.nio.channels.DatagramChannel;

public class Udplib {
  private int port;
  private DatagramChannel handle;

  /**
   * Constructor
   *
   * @param port udp port to use
   */
  public Udplib(int port) throws IOException {
    this.port = port;
    //  Create UDP socket
    handle = DatagramChannel.open();

    //  Ask operating system to let us do broadcasts from socket
    handle.setOption(StandardSocketOptions.SO_BROADCAST, true);
    handle.configureBlocking(false);

    //  Allow multiple processes to bind to socket; incoming
    //  messages will come to each process
    handle.setOption(StandardSocketOptions.SO_REUSEADDR, true);

    handle.bind(new InetSocketAddress(port));
  }

  /**
   * Returns UDP socket handle
   */
  public DatagramChannel handle() {
    return handle;
  }

  /**
   * Send message using UDP broadcast
   */
  public void send(ByteBuffer buffer) throws IOException {
    handle.send(buffer, new InetSocketAddress("255.255.255.255", port));
  }

  /**
   * Receive message from UDP broadcast
   *
   * @return source of the datagram or null if there isn't any datagram to receive
   */
  public SocketAddress recv(ByteBuffer buffer) throws IOException {
    return handle.receive(buffer);
  }

  /**
   * Destructor
   */
  public void destroy() throws IOException {
    handle.close();
  }
}
